src/ctap2/commands/mod.rs

```diff
@@ -92,66 +95,179 @@ pub trait RequestCtap2: fmt::Debug {
+
+    /// If Ok() is returned, the request can be sent to the device
+    /// If Err() is returned, either cancel the operation or ask the user for a PIN
+    fn determine_pin_uv_auth<D: FidoDevice>(
+        &mut self,
+        dev: &mut D,
+        skip_uv: bool,
+    ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        // CTAP1/U2F-only devices do not support PinUvAuth, so we skip it
         if !dev.supports_ctap2() {
             self.set_pin_auth(None);
-            return Ok(());
+            return Ok(PinUvAuthResult::DeviceIsCtap1);
         }
 
-        let client_data_hash = self.client_data_hash();
-        let pin_auth = calculate_pin_auth(dev, &client_data_hash, self.pin())
+        self.determine_pin_uv_auth_ctap_2_0(dev, skip_uv)
+    }
+
+    /// CTAP 2.0-only version:
+    /// "Getting pinUvAuthToken using getPinToken (superseded)"
+    fn determine_pin_uv_auth_ctap_2_0<D: FidoDevice>(
+        &mut self,
+        dev: &mut D,
+        skip_uv: bool,
+    ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        let info = dev
+            .get_authenticator_info()
+            .ok_or(AuthenticatorError::Platform)?;
+
+        if self.pin().is_none() {
+            // Only use UV, if the device supports it and we don't skip it
+            // which happens as a fallback, if UV-usage failed too many times
+            let supports_uv = !skip_uv && info.options.user_verification == Some(true);
+            let supports_pin = info.options.client_pin.is_some();
+            let pin_configured = info.options.client_pin == Some(true);
+
+            // If the device supports internal user-verification (e.g. fingerprints),
+            // skip PIN-stuff
+            if supports_uv {
+                return Ok(PinUvAuthResult::UsingInternalUv);
+            }
+
+            // Device does not support any auth-method
+            if !supports_pin && !supports_uv {
+                // We'll send it to the device anyways, and let it error out (or magically work)
+                return Ok(PinUvAuthResult::NoAuthTypeSupported);
+            }
+
+            // Device supports PIN: Ask user for it.
+            if !supports_uv && pin_configured {
+                return Err(AuthenticatorError::PinError(PinError::PinRequired));
+            }
+        }
+
+        let pin_auth = dev
+            .get_pin_token(&self.client_data_hash(), self.pin())
             .map_err(|e| repackage_pin_errors(dev, e))?;
+
+        // CTAP 2.0 spec is a bit vague here, but CTAP 2.1 is very specific, that the request
+        // should either include pinAuth OR uv=true, but not both at the same time.
+        // Do not set user_verification, if pinAuth is provided
+        if pin_auth.is_some() {
+            self.unset_uv_option();
+        }
         self.set_pin_auth(pin_auth);
-        Ok(())
+        Ok(PinUvAuthResult::SuccessGetPinToken)
+    }
+
+    fn handle_auth_command_errors<D: FidoDevice>(
```

Comment:
This does not really "unify the cases" as we're now just pushing the dependence on Self::command() into this function. I'm sorry to go back and forth on this minor point, but I think we should do the flat list of cases instead.

Reply:
So reverting this change and doing what we did before? Or something else entirely?
"Avoiding duplicate code, if possible" and "having a flat list" unfortunately are at odds in this case.

Reply:
Maybe it will be easier for me to just submit a follow up patch with what I'm imagining. This is not as clear as I'd like it to be, but it does appear to be correct.